### PR TITLE
MODSOURMAN-1212: Update MARC bib-instance default mapping for subject source and subject type

### PR DESCRIPTION
**Pull request by VRohach**
## Purpose
 In order to support further subject heading granularity, we need to update the default MARC-Instance map to populate the new fields for Subject source and Subject type.

## Approach
Default mapping rules improved to be able to provide mapping for Subject Source and Subject Type based specific conditions from JIRA-ticket.
rules extended
additional schemas added
mapping parameteres extended
tests fixed+added
normalization functions added

## Learning
JIRA: https://folio-org.atlassian.net/browse/MODSOURMAN-1212

### CODE REVIEW

**Comment by VRohach** [comment]
After this merge, the another PR should be merged: https://github.com/folio-org/mod-source-record-manager/pull/926

File: src/main/java/org/folio/processing/mapping/defaultmapper/processor/functions/NormalizationFunction.java [hunk 1]
```diff
@@ -422,6 +424,42 @@ public String apply(RuleExecutionContext context) {
...
+      String typeName = context.getRuleParameter().getString(NAME_PARAMETER);
+      List<SubjectSource> subjectSources = context.getMappingParameters().getSubjectSources();
+      if (subjectSources == null || typeName == null) {
+        return STUB_FIELD_TYPE_ID;
+      }
+      return subjectSources.stream()
+        .filter(identifierType -> identifierType.getName().trim().equalsIgnoreCase(typeName))
+        .findFirst()
+        .map(SubjectSource::getId)
+        .orElse(STUB_FIELD_TYPE_ID);
```
Comment (KaterynaSenchenko) [inline src/main/java/org/folio/processing/mapping/defaultmapper/processor/functions/NormalizationFunction.java:441]: Is this stub id corresponds to the default subject source?

File: src/test/java/org/folio/processing/mapping/InstanceMappingTest.java [hunk 2]
```diff
@@ -461,6 +471,84 @@ public void testMarcToInstanceWithRepeatableISBN() throws IOException {
...
+    assertFalse(mappedInstances.isEmpty());
+    assertEquals(1, mappedInstances.size());
+/*    Set<Subject> subjects = mappedInstances.get(0).getSubjects();
+    assertEquals(6, subjects.size());
+    IntStream.range(0, expectedResults.size()).forEach(index -> {
+      Map.Entry<String, String> expected = expectedResults.get(index);
+      Identifier actual = subjects.get(index);
+      assertEquals(expected.getValue(), actual.getIdentifierTypeId());
+      assertEquals(expected.getKey(), actual.getValue());
+    });*/
```
Comment (RuslanLavrov) [inline src/test/java/org/folio/processing/mapping/InstanceMappingTest.java:537]:
```suggestion
```

File: src/test/java/org/folio/processing/mapping/InstanceMappingTest.java [hunk 3]
```diff
@@ -461,6 +471,84 @@ public void testMarcToInstanceWithRepeatableISBN() throws IOException {
     });
   }
 
+  @Test
+  public void testMarcToInstanceWithRepeatableSubects() throws IOException {
+    final String FIRST_LIBRARY_SOURCE_ID = "e894d0dc-621d-4b1d-98f6-6f7120eb0d40";
+    final String SECOND_LIBRARY_SOURCE_ID = "e894d0dc-621d-4b1d-98f6-6f7120eb0d40";
```
Comment (RuslanLavrov) [inline src/test/java/org/folio/processing/mapping/InstanceMappingTest.java:477]:
```suggestion
```
looks like this variable contains the same UUID as the previous one and is never used in the test

File: src/test/java/org/folio/processing/mapping/InstanceMappingTest.java [hunk 4]
```diff
@@ -461,6 +471,84 @@ public void testMarcToInstanceWithRepeatableISBN() throws IOException {
...
+    final String EIGHTH_LIBRARY_SOURCE_ID = "e894d0dc-621d-4b1d-98f6-6f7120eb0d47";
+
+    final String FIRST_SUBJECT_TYPE_ID = "d6488f88-1e74-40ce-81b5-b19a928ff5b1";
+    final String SECOND_SUBJECT_TYPE_ID = "d6488f88-1e74-40ce-81b5-b19a928ff5b2";
+    final String THIRD_SUBJECT_TYPE_ID = "d6488f88-1e74-40ce-81b5-b19a928ff5b3";
+    final String FOURTH_SUBJECT_TYPE_ID = "d6488f88-1e74-40ce-81b5-b19a928ff5b4";
+    final String FIFTH_SUBJECT_TYPE_ID = "d6488f88-1e74-40ce-81b5-b19a928ff5b5";
+    final String SIXTH_SUBJECT_TYPE_ID = "d6488f88-1e74-40ce-81b5-b19a928ff5b6";
+    final String SEVENTH_SUBJECT_TYPE_ID = "d6488f88-1e74-40ce-81b5-b19a928ff5b7";
+    final String EIGHTH_SUBJECT_TYPE_ID = "d6488f88-1e74-40ce-81b5-b19a928ff511";
```
[the remark this answers is not in the document]
Comment (VRohach) [reply src/test/java/org/folio/processing/mapping/InstanceMappingTest.java:492]: @RuslanLavrov  Thank you! You found a potential bug.

File: src/test/java/org/folio/processing/mapping/InstanceMappingTest.java (hunk 2 above)
Comment (VRohach) [reply src/test/java/org/folio/processing/mapping/InstanceMappingTest.java:537]: @RuslanLavrov  Thanks.

File: src/test/java/org/folio/processing/mapping/InstanceMappingTest.java (hunk 3 above)
Comment (VRohach) [reply src/test/java/org/folio/processing/mapping/InstanceMappingTest.java:477]: @RuslanLavrov  Thank you! I changed this logic.

File: src/test/java/org/folio/processing/mapping/InstanceMappingTest.java [hunk 5]
```diff
@@ -461,6 +471,84 @@ public void testMarcToInstanceWithRepeatableISBN() throws IOException {
...
+      JsonObject marc = new JsonObject(os.toString());
+      Instance instance = mapper.mapRecord(marc, new MappingParameters().withSubjectSources(subjectSources).withSubjectTypes(subjectTypes), mappingRules);
+      mappedInstances.add(instance);
+      Validator validator = factory.getValidator();
+      Set<ConstraintViolation<Instance>> violations = validator.validate(instance);
+      assertTrue(violations.isEmpty());
+    }
+    assertFalse(mappedInstances.isEmpty());
+    assertEquals(1, mappedInstances.size());
+/*    Set<Subject> subjects = mappedInstances.get(0).getSubjects();
```
[the remark this answers is not in the document]
Comment (VRohach) [reply src/test/java/org/folio/processing/mapping/InstanceMappingTest.java:530]: @JavokhirAbdullayev  Thank you! removed.

**Review by RuslanLavrov — APPROVED** [review]

**Review by KaterynaSenchenko — APPROVED** [review]

File: src/main/java/org/folio/processing/mapping/defaultmapper/processor/functions/NormalizationFunction.java (hunk 1 above)
Comment (VRohach) [reply src/main/java/org/folio/processing/mapping/defaultmapper/processor/functions/NormalizationFunction.java]: @KaterynaSenchenko Good point! After verifying with PO I changed this logic and set it as empty if there will be NO MATCH.

**Pull request merged**